### PR TITLE
Additional handles in generated code

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR addresses #68. 
It sets the line numbers for autogenerated statements to the line number of the corresponding implicit or explicit specialization declaration in the Q# code, and adds an attribute for each specialization to the callable class that contains the source file where it has been specified as well as the start and end line of the specialization declaration.
The line numbers are 0-based and the end line number is set to -1 if it is the last specialization declaration in the file.

### CODE REVIEW

File: src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs [hunk 1]
```diff
@@ -2303,6 +2304,10 @@ namespace N1
         |> testOneClass randomAbstractOperation
 
         """
+    [SpecializationRangeAttribute("%%%", "Body", 107, 112)]
```
Comment (anpaz-msft) [inline src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs]: nit: we could simple use [SepecializationRange()]

File: src/Simulation/Core/Attributes.cs [hunk 2]
```diff
@@ -6,6 +6,23 @@
 
 namespace Microsoft.Quantum.Simulation.Core
 {
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
+    public class SpecializationRangeAttribute : Attribute
```
Comment (anpaz-msft) [inline src/Simulation/Core/Attributes.cs]: nit: would `SpecializationSourceLocationAttribute` or just `SourceLocation` be a better name?

File: src/Simulation/Core/Attributes.cs [hunk 3]
```diff
@@ -6,6 +6,23 @@
...
+    { 
+        public string SourceFile { get; }
+        public string SpecializationKind { get; }
```
Comment (vadym-kl) [inline src/Simulation/Core/Attributes.cs]: Sorry for not pointing out this in earlier discussion, but there is `enum`
https://github.com/microsoft/qsharp-runtime/blob/ebb4b2a9de0110ab45a6e38b38cfbd664759483f/src/Simulation/Core/OperationFunctor.cs#L9 
that might be useful here.

File: src/Simulation/Core/Attributes.cs (hunk 2 above)
Comment (bettinaheim) [reply src/Simulation/Core/Attributes.cs]: Yes, renamed it to SourceLocationAttribute.

File: src/Simulation/Core/Attributes.cs (hunk 3 above)
Comment (bettinaheim) [reply src/Simulation/Core/Attributes.cs]: changed.

**Pull request merged**